### PR TITLE
Rule no watch

### LINKED ISSUE #48: Rule: no-watch
Watch have last  [computation priority](https://effector.dev/docs/advanced-guide/computation-priority).

Frequent use of the watch can lead to races that are very difficult to debug. Often, the use of the watch can be avoided (by using `target`, for example).

### PR DESCRIPTION
**Pull request by binjospookie**
close #48

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,3 +48,4 @@ To configure individual rules:
 - [effector/no-useless-methods](/rules/no-useless-methods/no-useless-methods.md)
 - [effector/no-ambiguity-target](/rules/no-ambiguity-target/no-ambiguity-target.md)
 - [effector/prefer-sample-over-forward-with-mapping](/rules/prefer-sample-over-forward-with-mapping/prefer-sample-over-forward-with-mapping.md)
+- [effector/@typescript-no-watch](/rules/@typescript-no-watch/@typescript-no-watch.md)
```
Comment (igorkamyshev) [inline README.md:51]: Maybe we ought to omit `typescript` from the name? What if in future someone'll find a way to improve it in JS?

File: rules/@typescript-no-watch/@typescript-no-watch.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        if (methodName !== "watch") {
+          return;
+        }
+
+        const object = traverseNestedObjectNode(node.callee?.object);
+        const objectName = object?.name;
+        const calleeName = object?.callee?.name;
+
+        if (!objectName && !["sample", "guard"].includes(calleeName)) {
```
Comment (igorkamyshev) [inline rules/@typescript-no-watch/@typescript-no-watch.js:34]: I do not understand the second part of check. I thought, we can check the return type of operator and apply rule. E.g. `sample` returns `Event`, so let's check it.

In my opinion, in this case, we can cover more cases — patronum-methods, for example.

File: rules/@typescript-no-watch/@typescript-no-watch.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
+# effector/@typescript-no-watch
+
+Method `.watch` leads to imperative code. Try replacing it with operators (`forward`, `sample`, etc) or use the `target` parameter of the operators.
+
+```ts
```
Comment (igorkamyshev) [inline rules/@typescript-no-watch/@typescript-no-watch.md:5]: Such number of examples leads to poor documentation reading experience. Could we replace it with a few amount of the most impressive cases?

File: rules/@typescript-no-watch/@typescript-no-watch.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
+# effector/@typescript-no-watch
+
```
Comment (igorkamyshev) [inline rules/@typescript-no-watch/@typescript-no-watch.md:2]: In my mind, we should warn a user about issues with JS-way in this rule right here.

File: rules/@typescript-no-watch/examples/correct.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const event = {
```
Comment (igorkamyshev) [inline rules/@typescript-no-watch/examples/correct.ts:1]: Let's add some smoke cases to correct test. E.g. `sample({ ... })`, `forward({...}), etc.

**Review by igorkamyshev — CHANGES_REQUESTED** [review]
Thanks for your contribution, it's a great help 💙

I've left some questions about the code and the documentation.

In general, I like it. I think we can release this rule in 0.4.0 on the next week.

File: README.md (hunk 1 above)
Comment (binjospookie) [reply README.md:51]: I'll fix it after I make the rest of the changes.

**Comment by binjospookie** [comment]
Made all necessary changes.

File: rules/no-watch/no-watch.ts.test.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  },
+});
+
+const readExampleForTheRule = (name) => ({
+  code: readExample(__dirname, name),
+  filename: join(__dirname, "examples", name),
+});
+
+ruleTester.run("effector/no-watch.ts.test", rule, {
+  valid: ["correct.ts"].map(readExampleForTheRule),
```
Comment (igorkamyshev) [inline rules/no-watch/no-watch.ts.test.js:24]: It's still noting for correct cases like `forward({...})`, factories, etc. 🥺

File: rules/no-watch/no-watch.ts.test.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+  invalid: [
+    ...["incorrect-with-effect.ts"]
+      .map(readExampleForTheRule)
+      .map((result) => ({
+        ...result,
+        errors: [
```
Comment (igorkamyshev) [inline rules/no-watch/no-watch.ts.test.js:31]: This pattern leads to hard debugging. Could you split incorrect test-cases and put it by files with only one error?

**Review by igorkamyshev — COMMENTED** [review]
Great job!

I think, we are ready to merge it. I've left some comments about test-cases. Once you fix it, I'll merge PR.

**Comment by binjospookie** [comment]
Nice. В ближайшие дни я буду за городом. До выходных закончу с этим pr

**Comment by binjospookie** [comment]
Nice. I'll try to fix it asap 🙂

File: rules/no-watch/no-watch.ts.test.js (hunk 6 above)
Comment (binjospookie) [reply rules/no-watch/no-watch.ts.test.js:24]: What do you mean? I added cases for guard, sample, effects and etc inside correct.ts

File: rules/no-watch/no-watch.ts.test.js (hunk 7 above)
Comment (binjospookie) [reply rules/no-watch/no-watch.ts.test.js:31]: Do you mean that for every incorrect case there should be a separate file?
1 error per file

For example: effects has 5 cases => 5 files

Am I right?

Comment (igorkamyshev) [reply rules/no-watch/no-watch.ts.test.js:31]: Hm, in this case up to you. I think, it's easier to debug in case of broken tests, but I can be wrong here.

Comment (binjospookie) [reply rules/no-watch/no-watch.ts.test.js:31]: You right. I choose a solution in the middle. No more than 2 errors per case.

**Comment by binjospookie** [comment]
I'm ready to merge🙃

**Comment by igorkamyshev** [comment]
Thanks! Will be released in a week with `0.4.0`.

**Comment by binjospookie** [comment]
Thank u too!

**Pull request merged**